examples/fluids/src/petsc_ops.c

```diff
@@ -344,4 +398,46 @@ PetscErrorCode MatGetDiag_Ceed(Mat A, Vec D) {
   PetscFunctionReturn(0);
 };
 
+/**
+ * @brief Create PETSc MatShell object for the corresponding OperatorApplyContext
+ *
+ * @param ctx Context that does the action of the operator
+ * @param mat MatShell for the operator
+ */
+PetscErrorCode CreateMatShell_Ceed(OperatorApplyContext ctx, Mat *mat) {
+  MPI_Comm comm_x = PetscObjectComm((PetscObject)(ctx->dm_x));
+  MPI_Comm comm_y = PetscObjectComm((PetscObject)(ctx->dm_y));
+  PetscInt X_loc_size, X_size, Y_size, Y_loc_size;
+  VecType  X_vec_type, Y_vec_type;
+  Vec      X, Y;
+
+  PetscFunctionBeginUser;
+
+  PetscCheck(comm_x == comm_y, PETSC_COMM_WORLD, PETSC_ERR_ARG_NOTSAMECOMM, "Input and output comm must be the same");
+
+  PetscCall(DMGetGlobalVector(ctx->dm_x, &X));
+  PetscCall(VecGetLocalSize(X, &X_loc_size));
+  PetscCall(VecGetSize(X, &X_size));
+  PetscCall(VecGetType(X, &X_vec_type));
+  PetscCall(DMRestoreGlobalVector(ctx->dm_x, &X));
+
+  PetscCall(DMGetGlobalVector(ctx->dm_y, &Y));
+  PetscCall(VecGetLocalSize(Y, &Y_loc_size));
+  PetscCall(VecGetSize(Y, &Y_size));
+  PetscCall(VecGetType(Y, &Y_vec_type));
+  PetscCall(DMRestoreGlobalVector(ctx->dm_y, &Y));
+
+  PetscCall(MatCreateShell(comm_x, Y_loc_size, X_loc_size, Y_size, X_size, ctx, mat));
+  PetscCall(MatShellSetContextDestroy(*mat, (PetscErrorCode(*)(void *))OperatorApplyContextDestroy));
+  PetscCall(MatShellSetOperation(*mat, MATOP_MULT, (void (*)(void))MatMult_Ceed));
+  PetscCall(MatShellSetOperation(*mat, MATOP_GET_DIAGONAL, (void (*)(void))MatGetDiag_Ceed));
+
+  // TODO Is setting the vectype necessary? IJacobian_NS doesn't...
+  PetscCheck(X_vec_type == Y_vec_type, PETSC_COMM_WORLD, PETSC_ERR_ARG_NOTSAMETYPE, "Vec_type of ctx->dm_x (%s) and ctx->dm_y (%s) must be the same",
+             X_vec_type, Y_vec_type);
+  PetscCall(MatShellSetVecType(*mat, X_vec_type));
```

Comment:
Not sure about the usefulness/necessity of this. `MatShellSetVecType` only appears to affect `MatGetVecs`, which we don't use, so possibly a check for nothing.

Reply:
I agree its probably not needed but keeping it is probably not bad either.